packages/langium/src/utils/ast-util.ts

```diff
@@ -194,3 +195,52 @@ export function findLocalReferences(targetNode: AstNode, lookup = getDocument(ta
     });
     return stream(refs);
 }
+
+/**
+ * Creates a deep copy of the specified AST node.
+ * The resulting copy will only contain semantically relevant information, such as the `$type` property and AST properties.
+ *
+ * References are copied without resolved cross reference. The specified linker is used to resolve them when necessary.
+ */
+export function copyAstNode<T extends AstNode = AstNode>(node: T, linker: Linker): T {
```

Comment:
Proposal: accept a `buildReference` function instead of the whole linker (makes it clearer what the state of the references is in the copy).